# HG changeset patch
# User Vitaliy Filippov <vitali@st-filippov.office.custis.ru>
# Date 1270052406 -14400
Create branch for patch 000-debug-categorized-specialrecentchanges

diff -r 2f012c39365a -r 017ab49f652c includes/specials/SpecialRecentchanges.php
--- includes/specials/SpecialRecentchanges.php
+++ includes/specials/SpecialRecentchanges.php
@@ -66,7 +66,7 @@
 	public function feedSetup() {
 		global $wgFeedLimit, $wgRequest;
 		$opts = $this->getDefaultOptions();
-		$opts->fetchValuesFromRequest( $wgRequest, array( 'days', 'limit', 'hideminor' ) );
+		$opts->fetchValuesFromRequest( $wgRequest );
 		$opts->validateIntBounds( 'limit', 0, $wgFeedLimit );
 		return $opts;
 	}
@@ -264,6 +264,7 @@
 	 */
 	public function doMainQuery( $conds, $opts ) {
 		global $wgUser;
+		global $wgAllowCategorizedRecentChanges;
 
 		$tables = array( 'recentchanges' );
 		$join_conds = array();
@@ -274,11 +275,31 @@
 		$namespace = $opts['namespace'];
 		$invert = $opts['invert'];
 
+		$categories = trim( $opts['categories'], " \t\n\r\0\x0B|" );
+		// JOIN on categories
+		if( $wgAllowCategorizedRecentChanges && $categories &&
+		    ( $categories = preg_split( '/[\s\|]*\|[\s\|]*/', $categories ) ) )
+		{
+			foreach( $categories as &$cat )
+				$cat = str_replace( ' ', '_', $cat );
+			$tables[] = 'page';
+			$join_conds['page'] = array( 'INNER JOIN', array( 'page_title=rc_title', 'page_namespace=rc_namespace' ) );
+			if( $opts['categories_any'] )
+				$conds[] = "EXISTS (".$dbr->selectSQLText( 'categorylinks', '*', array( 'cl_from=page_id', 'cl_to' => $categories ) ).")";
+			else
+			{
+				foreach( $categories as $i => $cat )
+				{
+					$tables[] = "`categorylinks` cl$i";
+					$join_conds["`categorylinks` cl$i"] = array( "INNER JOIN", array( "cl$i.cl_from=page_id", "cl$i.cl_to" => $cat ) );
+				}
+			}
+		}
+
 		// JOIN on watchlist for users
 		if( $uid ) {
 			$tables[] = 'watchlist';
-			$join_conds = array( 'watchlist' => array('LEFT JOIN',
-				"wl_user={$uid} AND wl_title=rc_title AND wl_namespace=rc_namespace") );
+			$join_conds['watchlist'] = array('LEFT JOIN', array('wl_user' => $uid, 'wl_title=rc_title', 'wl_namespace=rc_namespace'));
 		}
 
 		wfRunHooks('SpecialRecentChangesQuery', array( &$conds, &$tables, &$join_conds, $opts ) );
@@ -322,7 +343,6 @@
 	 */
 	public function webOutput( $rows, $opts ) {
 		global $wgOut, $wgUser, $wgRCShowWatchingUsers, $wgShowUpdatedMarker;
-		global $wgAllowCategorizedRecentChanges;
 
 		$limit = $opts['limit'];
 
@@ -333,10 +353,7 @@
 
 		// And now for the content
 		$wgOut->setSyndicated( true );
-
-		if( $wgAllowCategorizedRecentChanges ) {
-			$this->filterByCategories( $rows, $opts );
-		}
+		$wgOut->setFeedAppendQuery( http_build_query( $opts->getAllValues() ) );
 
 		$showWatcherCount = $wgRCShowWatchingUsers && $wgUser->getOption( 'shownumberswatching' );
 		$watcherCache = array();
@@ -392,7 +409,7 @@
 
 		$defaults = $opts->getAllValues();
 		$nondefaults = $opts->getChangedValues();
-		$opts->consumeValues( array( 'namespace', 'invert' ) );
+		$opts->consumeValues( array( 'namespace', 'invert', 'categories', 'categories_any' ) );
 
 		$panel = array();
 		$panel[] = $this->optionsPanel( $defaults, $nondefaults );
@@ -529,6 +546,7 @@
 		# Filter articles
 		$articles = array();
 		$a2r = array();
+		$newrows = array();
 		foreach( $rows AS $k => $r ) {
 			$nt = Title::makeTitle( $r->rc_namespace, $r->rc_title );
 			$id = $nt->getArticleID();
@@ -539,8 +557,10 @@
 			if( !isset($a2r[$id]) ) {
 				$a2r[$id] = array();
 			}
+			$newrows[$k] = $r;
 			$a2r[$id][] = $k;
 		}
+		$rows = $newrows;
 
 		# Shortcut?
 		if( !count($articles) || !count($cats) )
